include/inexor/vulkan-renderer/gpu_memory_buffer.hpp

```diff
@@ -1,29 +1,82 @@
 ﻿#pragma once
 
-#include <vma/vma_usage.h>
+#include "vma/vk_mem_alloc.h"
+
+#include <spdlog/spdlog.h>
+
+#include <vulkan/vulkan.h>
 
-#include <cstddef>
 #include <string>
 
 namespace inexor::vulkan_renderer {
 
-/// @brief An abstraction class for the creation of staging buffers.
-/// @note We can't add a std::mutex in here because that would force Buffer to be uncopyable!
-struct Buffer {
-    // TODO: Inherit to make access unified in syntax!
-    // TODO: Map/Unmap memory.
-
+class GPUMemoryBuffer {
+protected:
     std::string name;
 
+    VkDevice device;
+    VmaAllocator vma_allocator;
     VkBuffer buffer = VK_NULL_HANDLE;
-
     VmaAllocation allocation = VK_NULL_HANDLE;
-
     VmaAllocationInfo allocation_info = {};
-
     VkBufferCreateInfo create_info = {};
-
     VmaAllocationCreateInfo allocation_create_info = {};
+
+public:
+    /// Delete the copy constructor so gpu memory buffers are move-only objects.
+    GPUMemoryBuffer(const GPUMemoryBuffer &) = delete;
+    GPUMemoryBuffer(GPUMemoryBuffer &&buffer) noexcept;
+
+    /// Delete the copy assignment operator so gpu memory buffers are move-only objects.
+    GPUMemoryBuffer &operator=(const GPUMemoryBuffer &) = delete;
+    GPUMemoryBuffer &operator=(GPUMemoryBuffer &&) noexcept = default;
+
+    /// @brief Creates a new GPU memory buffer.
+    /// @param device [in] The Vulkan device from which the buffer will be created.
+    /// @param vma_allocator [in] The Vulkan Memory Allocator library handle.
+    /// @param name [in] The internal name of the buffer.
+    /// @param size [in] The size of the buffer in bytes.
+    /// @param buffer_usage [in] The Vulkan buffer usage flags.
+    /// @param memory_usage [in] The Vulkan Memory Allocator library's memory usage flags.
+    GPUMemoryBuffer(const VkDevice &device, const VmaAllocator &vma_allocator, std::string &name, const VkDeviceSize &size,
```

Comment:
`name` should be const.

Reply:
Or if it can't be, we should pass it by value and move it.

Reply:
It can't be const since VMA doesn't accept pointer to const:

```
#if VMA_RECORDING_ENABLED
    allocation_create_info.flags = VMA_ALLOCATION_CREATE_MAPPED_BIT | VMA_ALLOCATION_CREATE_USER_DATA_COPY_STRING_BIT;
    allocation_create_info.pUserData = name.data();
#else
```

Reply:
You don't pass the parameter to the function. You pass the pointer to the `name` member, **THATS** why i suggested to use **always** `this->` if you access a class member ;)

Reply:
done.